### ISSUE TITLE
Use QueryBuilder functionality

### ISSUE THREAD

**Issue by pleerock** [MaikelH/db-access-perf-tests#2]
There are lot of things TypeORM does under the hood in `save` method, lot of useful features that take extra time which you can't find in sequalize. Please consider this fact in your tests. Also, if you want the most effecient way you need to use `QueryBuilder` which also allows you to perform bulk insertions:

```ts
await connection.createQueryBuilder()
     .insert()
     .into(User)
     .values(users)
     .execute()
```

shortcut of which is newly added `insert` method in `Repository` / `EntityManager`:

```ts
await connection.manager.insert(users);
```

**Comment by MaikelH** [comment]
I wonder how you found this repository :smile:. This repository is a Work in Progress currently. We did this because we noticed significant slowdowns when processing thousands of rows using TypeORM. We are still doing some research into to the problem, but it seems that in certain cases TypeORM running on node 6 has `O(n²)` runtime complexity. Which is something that is not desirable. 

> There are lot of things TypeORM does under the hood in save method, lot of useful features that take extra time which you can't find in sequalize.

Yes this definitely the case. Using an ORM always has a certain perfomance impact. This test should answer what the exact performance impact is and if that is acceptable for us.

We will definitely add a test using the querybuilder and see how it compares. I hope to have the results of all the tests in the coming days.

**Comment by pleerock** [comment]
> We will definitely add a test using the querybuilder and see how it compares. I hope to have the results of all the tests in the coming days.

Yes, please add them. TypeORM's goal is to provide both "lot of features+simplicity+convenience" and "performance", but unfortunately in most cases they are standing on opposite sides. That's why it provides you a choice: you can do it the most simple, convenient and with lot of extra features way using `save` method, or you can use query builder (or newly added `insert` method) and do it the most performant way.  

I'll help you to achieve performance that satisfies your needs.

**Comment by EvaLok** [comment]
@pleerock i think one of the bigger issues for us right now is how to address an `upsert ... returning *` scenario. the only way using entity TypeORM methods seems to work as desired is by making extra calls to the db, which is not great for performance. is there an upsert method of some kind that would fit our needs so that we don't have to write raw queries? obviously in this example there are only a few fields, but you can imagine how with a bunch of fields it gets pretty tedious to write. this is a really common use case when implementing an aggregate service.

example (i didn't actually run this, but hopefully you can get the idea..):
```typescript
import { Column, PrimaryColumn, Repository, UpdateDateColumn } from "typeorm";
import { Entity } from "typeorm/decorator/entity/Entity";

export class SomeServiceClass {
    constructor(
        private repo: Repository<SomeAggregate>
    ) {}

    public async handleSomeEventWithTypeORMMethods( event: ISomeEvent ): Promise<SomeAggregate>  {
        let model = await this.repo.findOne({
            id: event.id,
        });

        if ( ! model ) {
            model = new SomeAggregate();
        }

        model.valueToUpdate = event.valueToUpdate;

        await this.repo.persist(model);

        const updatedModel = await this.repo.findOne({
            id: event.id,
        });

        return updatedModel;
    }

    public async handleSomeEventWithRawQuery( event: ISomeEvent ): Promise<SomeAggregate>  {
        const results = await this.repo.query(`
            INSERT INTO
                some_table
            (
                "id",
                "valueToUpdate"
            )
                VALUES
            (
                $1,
                $2
            )

            ON CONFLICT( "id" )
                DO UPDATE
                    SET
                        "valueToUpdate" = $2
                    WHERE
                        "some_table"."id" = $1

            RETURNING *
        `, [
            event.id,
            event.valueToUpdate
        ]);

        return results[0];
    }
}


@Entity()
export class SomeAggregate {
    @PrimaryColumn("uuid")
    public id: string;

    @Column("text")
    public valueToUpdate ?: string;

    @Column("text")
    public anotherValue ?: string;

    @UpdateDateColumn({type: "timestamp with time zone"})
    public modifiedAt ?: Date;
}

export interface ISomeEvent {
    id: string;
    valueToUpdate: string;
}

```

**Comment by pleerock** [comment]
two queries instead of one is not quite a big deal :) But I think feature you are requested can be implemented as part of `InsertQueryBuilder` functionality, why didn't you try to open an issue?

**Comment by EvaLok** [comment]
twice the amount of queries can be a big deal, especially in the case of aggregates which are updated frequently and have a lot of records. we'll try to quantify exactly what the performance difference is and get back to you once we're sure it's a necessary feature. as to why no issue.. i think there is a fairly similar one already opened by someone(s) else:

https://github.com/typeorm/typeorm/issues/978

maybe in addition to a `InseryQueryBuilder` method, something like a `persistAndReturn`, `updateAndReturn` or so might be pretty nice too

**Comment by pleerock** [comment]
> typeorm/typeorm#978

yes, its about returning and if you read issue it says that currently `InsertQueryBuilder` supports returning functionality:

```ts
return connection
      .createQueryBuilder(ControlPlace, 'controlplace')
      .insert()
      .values(controlPlace)
      .returning(...)
      .execute();
```

**Comment by EvaLok** [comment]
@pleerock unless it's been resolved, someone mentioned that it doesn't update the auto-generated columns. the solution you suggested of re-loading the entity results in an extra call to the database.

also - does the query builder support upsert? the example that you gave will not work if some records are pre-existing and others are not. looking at the docs, i only see `select`, `insert`, `update`, `delete`

in the case of an aggregate (for example), we cannot know if values are pre-existing or not without a query to the database or some sort of cache system, both of these are less efficient (at least that would be my assumption.. we can test it of course..) and more work than a `insert.. on conflict do.. returning..` query

**Comment by pleerock** [comment]
> @pleerock unless it's been resolved, someone mentioned that it doesn't update the auto-generated columns

yes, it does not update create a update dates

> the solution you suggested of re-loading the entity results in an extra call to the database.

right, or ORM could do that for you, but you want to avoid extra queries, right? =)

> does the query builder support upsert?

no, thats why I told you to create a new github issue for a feature you are requesting, you can even implement it on your own, not a hard feature to implement once we decide about its design.

**Comment by EvaLok** [comment]
@pleerock made an issue:
https://github.com/typeorm/typeorm/issues/1090

as far as the ORM handling upserts, in my proposed solution databases which support the `returning` syntax (or variation) can handle the operation with a single call, so the ORM should be able to do it without more than one query being executed.
